File: iosApp/iosApp/SwiftUIDemo/Scenes/ScenesViewModel.swift
import Foundation
import ComposeApp

final class ScenesViewModel: ObservableObject {
    static let layerLow = "low"
    static let layerHigh = "high"

    @Published var fadeMs: Float = 2_000
    @Published var currentId: String?
    @Published var layerIds: [String] = []
    @Published var json = ""
    @Published var isPlaying = false
    @Published var status = ""

    let low: AudioScene
    let high: AudioScene
    let stack: AudioScene
    let lowHall: AudioScene

    private var engine: AudioEngine?
    private var player: AudioScenePlayer?
    private var attached: [AudioStream] = []
    private var callbacks: [AudioStreamCallbackImpl] = []
    private var phases: [String: Float] = [:]

    init() {
        low = AudioScene(
            id: "low",
            layers: [SceneLayer(id: Self.layerLow, gainDb: 0, effects: [])]
        )
        high = AudioScene(
            id: "high",
            layers: [SceneLayer(id: Self.layerHigh, gainDb: 0, effects: [])]
        )
        stack = AudioScene(
            id: "stack",
            layers: [
                SceneLayer(id: Self.layerLow, gainDb: 0, effects: []),
                SceneLayer(id: Self.layerHigh, gainDb: -6, effects: []),
            ]
        )
        let reverbParams = KotlinMutableDictionary<KotlinInt, KotlinFloat>()
        reverbParams[KotlinInt(int: ReverbParams.shared.ROOM_SIZE)] = KotlinFloat(float: 0.7)
        reverbParams[KotlinInt(int: ReverbParams.shared.WET_DRY_MIX)] = KotlinFloat(float: 0.4)
        lowHall = AudioScene(
            id: "low-hall",
            layers: [
                SceneLayer(
                    id: Self.layerLow,
                    gainDb: 0,
                    effects: [
                        SceneEffect(
                            type: .reverb,
                            // Map<Int, Float> is [KotlinInt: KotlinFloat]; Swift literals fail to type-check.
                            params: reverbParams as NSDictionary as! [KotlinInt: KotlinFloat]
                        ),
                    ]
                ),
            ]
        )
    }

    func hz(for layerId: String) throws -> Double {
        guard let hz = Self.frequency(for: layerId) else {
            throw ScenesError.unknownLayer(layerId)
        }
        return hz
    }

    func transition(to scene: AudioScene) {
        do {
            for layer in scene.layers {
                _ = try hz(for: layer.id)
            }
            let active = ensurePlayer()
            let source = SceneLayerSource { [unowned self] layer in
                self.openLayer(layer.id)
            }
            active.transitionTo(scene: scene, fadeMs: fadeMs, source: source)
            let current = active.current
            currentId = current?.id
            layerIds = current?.layers.map(\.id) ?? []
            json = current.map { AudioSceneJson.shared.encode(scene: $0) } ?? ""
            isPlaying = true
            status = ""
        } catch {
            let message = error.localizedDescription
            status = message.isEmpty ? "transition failed" : message
        }
    }

    func stop() {
        for stream in attached {
            DemoSession.shared.detach(stream)
        }
        attached.removeAll()
        player?.close()
        engine?.release()
        player = nil
        engine = nil
        callbacks.removeAll()
        phases.removeAll()
        isPlaying = false
        currentId = nil
        layerIds = []
        json = ""
        status = ""
    }

    deinit { stop() }

    private func ensurePlayer() -> AudioScenePlayer {
        if let player { return player }
        let newEngine = AudioEngine.companion.create()
        engine = newEngine
        let created = AudioScenePlayer(engine: newEngine)
        player = created
        return created
    }

    private static func frequency(for layerId: String) -> Double? {
        switch layerId {
        case layerLow: return 220
        case layerHigh: return 660
        default: return nil
        }
    }

    private func openLayer(_ layerId: String) -> AudioStream {
        let hz: Double
        if let value = Self.frequency(for: layerId) {
            hz = value
        } else {
            status = "Unknown layer id: \(layerId)"
            hz = 0
        }
        let sampleRate: Int32 = 48_000
        let eng: AudioEngine
        if let existing = engine {
            eng = existing
        } else {
            eng = AudioEngine.companion.create()
            engine = eng
        }
        let phaseKey = layerId
        if phases[phaseKey] == nil { phases[phaseKey] = 0 }
        let cb = AudioStreamCallbackImpl { [weak self] buffer, numFrames in
            guard let self else { return numFrames }
            let frames = numFrames.intValue
            let twoPi = Float.pi * 2
            let increment = twoPi * Float(hz) / Float(sampleRate)
            var phase = self.phases[phaseKey] ?? 0
            for i in 0..<frames {
                buffer.set(index: Int32(i), value: sinf(phase) * 0.2)
                phase += increment
                if phase > twoPi { phase -= twoPi }
            }
            self.phases[phaseKey] = phase
            return numFrames
        }
        callbacks.append(cb)
        let config = AudioStreamConfig(
            sampleRate: sampleRate,
            channelCount: 1,
            audioFormat: .pcmFloat,
            bufferCapacityInFrames: 0,
            performanceMode: .lowLatency,
            sharingMode: .shared,
            direction: .output,
            deviceId: nil
        )
        let stream = eng.openStream(config: config, callback: cb)
        DemoSession.shared.attach(stream)
        attached.append(stream)
        return stream
    }
}

private enum ScenesError: LocalizedError {
    case unknownLayer(String)

    var errorDescription: String? {
        switch self {
        case .unknownLayer(let layerId): return "Unknown layer id: \(layerId)"
        }
    }
}

private final class SceneLayerSource: NSObject, AudioScenePlayerLayerSource {
    private let open: (SceneLayer) -> AudioStream

    init(open: @escaping (SceneLayer) -> AudioStream) {
        self.open = open
    }

    func openStream(layer: SceneLayer) -> AudioStream {
        open(layer)
    }
}

File: demo/src/commonMain/kotlin/com/vectencia/klarinet/demo/ScenesViewModel.kt
package com.vectencia.klarinet.demo

import androidx.lifecycle.ViewModel
import com.vectencia.klarinet.AudioEngine
import com.vectencia.klarinet.AudioScene
import com.vectencia.klarinet.AudioSceneJson
import com.vectencia.klarinet.AudioScenePlayer
import com.vectencia.klarinet.AudioStream
import com.vectencia.klarinet.AudioStreamCallback
import com.vectencia.klarinet.AudioStreamConfig
import com.vectencia.klarinet.StreamDirection
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlin.math.sin

data class ScenesUiState(
    val fadeMs: Float = 2000f,
    val currentId: String? = null,
    val layerIds: List<String> = emptyList(),
    val json: String = "",
    val isPlaying: Boolean = false,
    val status: String = "",
)

sealed interface ScenesEvent {
    data class FadeMs(val value: Float) : ScenesEvent
    data class Select(val scene: AudioScene) : ScenesEvent
    data object Stop : ScenesEvent
}

class ScenesViewModel : ViewModel() {
    private val _uiState = MutableStateFlow(ScenesUiState())
    val uiState: StateFlow<ScenesUiState> = _uiState.asStateFlow()

    private var engine: AudioEngine? = null
    private var player: AudioScenePlayer? = null
    private val attached = mutableListOf<AudioStream>()
    private val phases = mutableMapOf<String, Float>()

    fun onEvent(event: ScenesEvent) {
        when (event) {
            is ScenesEvent.FadeMs -> _uiState.update { it.copy(fadeMs = event.value) }
            is ScenesEvent.Select -> transition(event.scene)
            ScenesEvent.Stop -> stop()
        }
    }

    private fun ensurePlayer(): AudioScenePlayer {
        val existing = player
        if (existing != null) return existing
        val newEngine = AudioEngine.create()
        engine = newEngine
        val created = AudioScenePlayer(newEngine)
        player = created
        return created
    }

    private fun transition(scene: AudioScene) {
        try {
            val active = ensurePlayer()
            active.transitionTo(scene, _uiState.value.fadeMs) { layer ->
                openLayer(layer.id)
            }
            val current = active.current
            _uiState.update {
                it.copy(
                    currentId = current?.id,
                    layerIds = current?.layers?.map { layer -> layer.id } ?: emptyList(),
                    json = current?.let { AudioSceneJson.encode(it) } ?: "",
                    isPlaying = true,
                    status = "",
                )
            }
        } catch (error: Exception) {
            _uiState.update { it.copy(status = error.message ?: "transition failed") }
        }
    }

    private fun openLayer(layerId: String): AudioStream {
        val hz = ScenePresets.hzFor(layerId)
        val sampleRate = 48_000
        val eng = engine ?: throw IllegalStateException("engine missing")
        val phaseKey = layerId
        phases.getOrPut(phaseKey) { 0f }
        val stream = eng.openStream(
            AudioStreamConfig(
                sampleRate = sampleRate,
                channelCount = 1,
                direction = StreamDirection.OUTPUT,
            ),
            object : AudioStreamCallback {
                override fun onAudioReady(buffer: FloatArray, numFrames: Int): Int {
                    val twoPi = (2.0 * kotlin.math.PI).toFloat()
                    val increment = (twoPi * hz / sampleRate).toFloat()
                    var phase = phases[phaseKey] ?: 0f
                    for (i in 0 until numFrames) {
                        buffer[i] = sin(phase.toDouble()).toFloat() * 0.2f
                        phase += increment
                        if (phase > twoPi) phase -= twoPi
                    }
                    phases[phaseKey] = phase
                    return numFrames
                }
            },
        )
        DemoSession.attach(stream)
        attached += stream
        return stream
    }

    private fun stop() {
        attached.toList().forEach { stream ->
            try { DemoSession.detach(stream) } catch (_: Exception) {}
        }
        attached.clear()
        try { player?.close() } catch (_: Exception) {}
        try { engine?.release() } catch (_: Exception) {}
        player = null
        engine = null
        phases.clear()
        _uiState.update {
            it.copy(isPlaying = false, currentId = null, layerIds = emptyList(), json = "", status = "")
        }
    }

    override fun onCleared() {
        stop()
    }
}
